### PR TITLE
feat(client): add `http_patch` method

### PR DESCRIPTION
**Pull request by JohnVillalovos**
In order to support some new API calls we need to support the HTTP `PATCH` method.

Closes: #2469

### CODE REVIEW

**Comment by JohnVillalovos** [comment]
@nejch Just wanted to get the ball rolling on this. Not sure how much should be in the PR.

Some unit tests are going to be wanted. Not sure about functional tests.

**Comment by lmilbaum** [comment]
How about placing the function in the backend?

**Comment by JohnVillalovos** [comment]
> How about placing the function in the backend?

Since I don't know much about the backend, can you explain why? I'm doing it like all the other `http_*` methods which currently exist.

**Comment by lmilbaum** [comment]
> > How about placing the function in the backend?
> 
> Since I don't know much about the backend, can you explain why? I'm doing it like all the other `http_*` methods which currently exist.

Moving the http_* functions to the backend has already started. It will save a future work of moving this one as well and there will be no need to change the function signature.

**Comment by JohnVillalovos** [comment]
> Moving the http_* functions to the backend has already started. It will save a future work of moving this one as well and there will be no need to change the function signature.

Sorry. I haven't looked into the back-end code. Can you explain how that will work? I don't really see much of anything in the back-end. I don't see any of the `http_*` methods there.

If all the other `http_*` methods need to be moved then I'm not sure how this one extra function is going to add much to the work required.

**Comment by lmilbaum** [comment]
> > Moving the http_* functions to the backend has already started. It will save a future work of moving this one as well and there will be no need to change the function signature.
> 
> Sorry. I haven't looked into the back-end code. Can you explain how that will work? I don't really see much of anything in the back-end. I don't see any of the `http_*` methods there. EDIT: I see `http_request`
> 
> If all the other `http_*` methods need to be moved then I'm not sure how this one extra function is going to add much to the work required.

You might want to take a look at #2435

**Comment by nejch** [comment]
@JohnVillalovos, I think this could be a `feat` as people can already use it for unimplemented endpoints after merge. 

> > > Moving the http_* functions to the backend has already started. It will save a future work of moving this one as well and there will be no need to change the function signature.
> > 
> > 
> > Sorry. I haven't looked into the back-end code. Can you explain how that will work? I don't really see much of anything in the back-end. I don't see any of the `http_*` methods there. EDIT: I see `http_request`
> > If all the other `http_*` methods need to be moved then I'm not sure how this one extra function is going to add much to the work required.
> 
> You might want to take a look at #2435

@lmilbaum in any case we will have these user-facing helper methods on the client, preferably backend-agnostic (https://python-gitlab.readthedocs.io/en/stable/api-levels.html#lower-level-api-http-methods).

I hope if we get this right only `http_request` and the associated typing will need a significant backend-specific implementation and the rest can wrap around it like we do now.

**Review by nejch — CHANGES_REQUESTED** [review]

**Comment by lmilbaum** [comment]
> @JohnVillalovos, I think this could be a `feat` as people can already use it for unimplemented endpoints after merge.
> 
> > > > Moving the http_* functions to the backend has already started. It will save a future work of moving this one as well and there will be no need to change the function signature.
> > > 
> > > 
> > > Sorry. I haven't looked into the back-end code. Can you explain how that will work? I don't really see much of anything in the back-end. I don't see any of the `http_*` methods there. EDIT: I see `http_request`
> > > If all the other `http_*` methods need to be moved then I'm not sure how this one extra function is going to add much to the work required.
> > 
> > 
> > You might want to take a look at #2435
> 
> @lmilbaum in any case we will have these user-facing helper methods on the client, preferably backend-agnostic (https://python-gitlab.readthedocs.io/en/stable/api-levels.html#lower-level-api-http-methods). They help users work around endpoints we haven't implemented in the library.
> 
> I hope if we get this right only `http_request` and the associated typing will need a significant backend-specific implementation and the rest can wrap around it like we do now.

I am not following what you are trying to say. In any case, maybe switching the return type in a similar way I've done in the PR #2435.

**Comment by nejch** [comment]
> I am not following what you are trying to say. In any case, maybe switching the return type in a similar way I've done in the PR #2435.

@lmilbaum since this PR is a separate concern, I think we can do that after, I can also rebase your PR with `http_patch` to avoid more rebase issues on your end. I think we can go ahead with this one just to unblock users who want to update protected branches in https://github.com/python-gitlab/python-gitlab/issues/2390.

**Review by nejch — APPROVED** [review]

**Pull request merged**
